ADR 108: ADR for extending E2E infrastructure so that we can check CometBFT's behaviour with respect to ABCI++ grammar. (#902)
cometbft/cometbft: docs/architecture/adr-108-e2e-abci++.md

```go
func (app *Application) logABCIRequest(req *abci.Request) error {
	if !app.cfg.ABCIRequestsLoggingEnabled {
		return nil
	}
	s, err := GetABCIRequestString(req)
	if err != nil {
		return err
	}
	app.logger.Info(s)
	return nil
}
```

`GetABCIRequestString(req)` is a new method that receives a request and returns its string representation. The implementation and tests for this function and the opposite function `GetABCIRequestFromString(req)`
that returns `abci.Request` from the string are provided in files `test/e2e/app/log.go` and `test/e2e/app/log_test.go`, respectively. To create a string representation of a request, we first marshal the request via `proto.Marshal()` method and then convert received bytes in the string using `base64.StdEncoding.EncodeToString()` method. In addition, we surround the new string with `abci-req` constants so that we can find lines with ABCI++ request more easily. The code of the method is below: 

```go
func GetABCIRequestString(req *abci.Request) (string, error) {
	b, err := proto.Marshal(req)
	if err != nil {
		return "", err
	}
	reqStr := base64.StdEncoding.EncodeToString(b)
	s := ABCI_REQ + reqStr + ABCI_REQ
	return s, nil
}
```

> *Note:* At the moment, we are not compressing the marshalled request before converting it to `base64` `string` because we are logging the empty requests that take at most 24B. However, if we decide to log the actual requests in the future, we might want to compress them. Based on a few tests, we observed that the size of a request can go up to 7KB.  

Reviewer: ```suggestion
*Note:* At the moment, we are not compressing the marshalled request before converting it to `base64` `string` because we are logging the empty requests that take at most 24 bytes. However, if we decide to log the actual requests in the future, we might want to compress them. Based on a few tests, we observed that the size of a request can go up to 7KB.  
```

If in the future we want to log another ABCI++ request type, we just need to do the same thing: 
create a corresponding `abci.Request` and log it via 
`app.logABCIRequest(r)`. 

### 2) Parsing the logs
We need a code that will take the logs from all nodes and collect the ABCI++ requests that were logged by the application. 

**Implementation**

This logic is implemented inside the `fetchABCIRequests(t *testing.T, nodeName string)` function that resides in `test/e2e/tests/e2e_test.go` file. This function does three things:
- Takes the output of a specific node in the testnet from the moment we launched the testnet until the function is called. The node name is passed as a function parameter. It uses the `docker-compose logs` and `grep nodeName` commands. 
- Parses the logs line by line and extracts the  `abci.Request`, if one exists. The request is received by forwarding each line to the `app.GetABCIRequestFromString(req)` method.

> - Returns the array of slices where each slice contains the set of `abci.Request`s logged on that node. Every time the crash happens, a new array element (new slice `[]*abci.Request`) will be created. We know a crash has happened because we log "Application started" every time the application starts. Specifically, we added this log inside `NewApplication()` function in `test/e2e/app/app.go` file. In the end, `fetchABCIRequests()` will return just one slice if the node did not experience any crashes and $n+1$ slices if there were crashes, $n$ being the number of crashes. The benefit of logging the requests in the previously described way is that now we can use `[]*abci.Request` to store ABCI++ requests of any type.

Reviewer: ```suggestion
- Returns the array of slices where each slice contains the set of `abci.Request`s logged on that node. Every time a crash happens, a new array element (new slice `[]*abci.Request`) will be created. We know a crash has happened because we log "Application started" every time the application starts. Specifically, we added this log inside `NewApplication()` function in `test/e2e/app/app.go` file. In the end, `fetchABCIRequests()` will return just one slice if the node did not experience any crashes and $n+1$ slices if there were crashes, $n$ being the number of crashes. The benefit of logging the requests in the previously described way is that now we can use `[]*abci.Request` to store ABCI++ requests of any type.
```

- `test/e2e/pkg/grammar/clean-start/lexer`,
- `test/e2e/pkg/grammar/clean-start/parser`,
- `test/e2e/pkg/grammar/clean-start/sppf`,
- `test/e2e/pkg/grammar/clean-start/token`,
- `test/e2e/pkg/grammar/recovery/lexer`,
- `test/e2e/pkg/grammar/recovery/parser`,
- `test/e2e/pkg/grammar/recoveryt/sppf`,
- `test/e2e/pkg/grammar/recovery/token`.

Apart from this auto-generated code, we implemented `GrammarChecker` abstraction
which knows how to use the generated parsers and lexers to verify whether a
specific execution (list of ABCI++ calls logged by the Application while the
testnet was running) respects the ABCI++ grammar. The implementation and tests 
for it are inside `test/e2e/pkg/grammar/checker.go` and 
`test/e2e/pkg/grammar/checker_test.go`, respectively. 

How the `GrammarChecker` works is demonstrated with the test `TestABCIGrammar`
implemented in `test/e2e/tests/abci_test.go` file. 

```go
func TestABCIGrammar(t *testing.T) {
	checker := grammar.NewGrammarChecker(grammar.DefaultConfig())
	testNode(t, func(t *testing.T, node e2e.Node) {
		if !node.Testnet.ABCITestsEnabled {
			return
		}
		reqs, err := fetchABCIRequests(t, node.Name)
		if err != nil {

> 			t.Error(fmt.Errorf("Collecting of ABCI requests failed: %w", err))

Reviewer: ```suggestion
			t.Error(fmt.Errorf("collecting of ABCI requests failed: %w", err))
```

logged by this node. Remember here that `fetchABCIRequests()` returns an array of slices(`[]*abci.Request`) where the slice 
with index 0 corresponds to the node's `CleanStart` execution, and each additional slice corresponds to the `Recovery` 
execution after a specific crash. Each node must have one `CleanStart` execution and the same number of `Recovery` executions 
as the number of crashes that happened on this node. If collecting was successful, the test checks whether each execution 
respects the ABCI++ 
grammar by calling `checker.Verify()` method. If `Verify` returns an error, the specific execution does not respect the 
grammar, and the test will fail. 

The tests are executed only if `ABCITestsEnabled` is set to `true`. This is done through the manifest file. Namely, if we 
want to test whether CometBFT respects ABCI++ grammar, we would need to enable these tests by adding `abci_tests_enabled = 
true` in the manifest file of a particular testnet (e.g. `networks/ci.toml`). This will automatically activate logging on the 
application side. 

The `Verify()` method is shown below. 
```go
func (g *GrammarChecker) Verify(reqs []*abci.Request, isCleanStart bool) (bool, error) {

> 	r := g.filterRequests(reqs)

Reviewer: Here, you are using `reqs` and validating its contents just after. Shouldn't you move this line after 196? (so that you first validate `reqs` and then use it)

Reviewer: .. or should line 194 use `r` rather than `reqs`?

	// This should not happen in our tests.
	if len(reqs) == 0 {

> 		return false, fmt.Errorf("Execution with no ABCI calls.")

Reviewer: ```suggestion
		return false, fmt.Errorf("execution with no ABCI calls.")
```

	}
	var errors []*Error
	execution := g.getExecutionString(r)
	if isCleanStart {
		errors = g.verifyCleanStart(execution)
	} else {
		errors = g.verifyRecovery(execution)
	}
	if errors == nil {
		return true, nil
	}
	return false, fmt.Errorf("%v\nFull execution:\n%v", g.combineErrors(errors, g.cfg.NumberOfErrorsToShow), g.addHeightNumbersToTheExecution(execution))
}
```

It receives a set of ABCI++ requests and a flag saying whether they represent a `CleanStart` execution or not and does the following things:
- Filter the requests by calling the method `filterRequests()`. This method will remove all the requests from the set that are not supported by the current version of the grammar. In addition, it will filter the last height by removing all ABCI++ requests after the 
last `Commit`. The function `fetchABCIRequests()` can be called in the middle of the height. As a result, the last height may be incomplete and 
classified as invalid, even if that is not the reality. The simple example here is that the last 

> request fetched via `fetchABCIRequests()` is `Decide`; however, `Commit` happens after 

Reviewer: By `Decide`, did you mean `EndBlock` (or `FinalizeBlock` in `v0.38.x`) ?

Author: Decide refers to FinalizeBlock: the terminal name in the grammar is Decide, while the value is FinalizeBlock 
(`decide = %s"<FinalizeBlock>"`). This may need to be changed, but it is like that also in the grammar on main and on v0.38.x branch.

Reviewer: I see. In `decide = %s"<FinalizeBlock>"`, technically, `decide` is a non-terminal that gets replaced by terminals "<FinalizeBlock>".
If you think this causes confusion, since we are not planning to backport this project (at least not beyond `v0.38.x`) maybe we can replace `decide` by `finalize`or `finalize-block` or something similar.

Author: Yes, we can, just we need to rename it everywhere I guess. In my ADR it is not the problem but not sure where else.

Reviewer: I don't think we need to replace all occurrences of the work decide (the concept is clear and understood), just the occurrences of the grammar's `decide` symbol. I quickly checked, and it looks like it's only 4 occurrences.

# ADR 108: E2E tests for CometBFT's behaviour in respect to ABCI++.

## Changelog
- 2023-08-08: Initial version (@nenadmilosevic95)

## Context

> We want to be able to test whether CommetBFT respects the ABCI++ grammar. To do this, we need to enhance the e2e tests infrastructure. Specifically, we plan to do three things:

Reviewer 2: ```suggestion
ABCI++ defines the interface between the application and CometBFT. A part of the specification is the ABCI++ grammar that describes the sequences of calls that the application can expect from CometBFT.
In order to demonstrate that CometBFT behaves as expected from the viewpoint of the application, we need to test whether CometBFT respects this ABCI++ grammar. To do this, we need to enhance the e2e tests infrastructure. Specifically, we plan to do three things:
```

> # ADR 108: E2E tests for CometBFT's behaviour in respect to ABCI++.

Reviewer: ```suggestion
# ADR 108: E2E tests for CometBFT's behaviour in respect to ABCI 1.0.
```

## Changelog
- 2023-08-08: Initial version (@nenadmilosevic95)

## Context

> ABCI++ defines the interface between the application and CometBFT. A part of the specification is the ABCI++ grammar that describes the sequences of calls that the application can expect from CometBFT.

Reviewer: ```suggestion
ABCI 1.0 defines the interface between the application and CometBFT. A part of the specification is the ABCI++ grammar that describes the sequences of calls that the application can expect from CometBFT.
```
Please, change ABCI++ to ABCI 1.0 all over. (ABCI 1.0 -> `PrepProp` and `ProcProp` but no vote extensions).
Whereas ABCI++ tends to be identifed with the whole lot (a.k.a. ABCI 2.0)

> ABCI++ defines the interface between the application and CometBFT. A part of the specification is the ABCI++ grammar that describes the sequences of calls that the application can expect from CometBFT.

Reviewer: Please add a link to the grammar's section in the spec

Author: Added.

In order to demonstrate that CometBFT behaves as expected from the viewpoint of the application, we need to test whether CometBFT respects this ABCI++ grammar. To do this, we need to enhance the e2e tests infrastructure. Specifically, we plan to do three things:
- Log every CometBFT's ABCI++ request during the execution.
- Parse the logs post-mortem and extract all ABCI++ requests.
- Check if the set of observed requests respects the ABCI++ grammar.

Issue: [353](https://github.com/cometbft/cometbft/issues/353).

> Current version does not support vote extensions. However, this is the next step. 

Reviewer: ```suggestion
Current version, ABCI 1.0, does not support vote extensions (ABCI 2.0). However, this is the next step. 
```

> 

Reviewer: ```suggestion
```

If in the future we want to log another ABCI++ request type, we just need to do the same thing: 
create a corresponding `abci.Request` and log it via 
`app.logABCIRequest(r)`. 

### 2) Parsing the logs
We need a code that will take the logs from all nodes and collect the ABCI++ requests that were logged by the application. 

**Implementation**

This logic is implemented inside the `fetchABCIRequests(t *testing.T, nodeName string)` function that resides in `test/e2e/tests/e2e_test.go` file. This function does three things:
- Takes the output of a specific node in the testnet from the moment we launched the testnet until the function is called. The node name is passed as a function parameter. It uses the `docker-compose logs` and `grep nodeName` commands. 
- Parses the logs line by line and extracts the  `abci.Request`, if one exists. The request is received by forwarding each line to the `app.GetABCIRequestFromString(req)` method.

> - Returns the array of slices where each slice contains the set of `abci.Request`s logged on that node. Every time a crash happens, a new array element (new slice `[]*abci.Request`) will be created. We know a crash has happened because we log "Application started" every time the application starts. Specifically, we added this log inside `NewApplication()` function in `test/e2e/app/app.go` file. In the end, `fetchABCIRequests()` will return just one slice if the node did not experience any crashes and $n+1$ slices if there were crashes, $n$ being the number of crashes. The benefit of logging the requests in the previously described way is that now we can use `[]*abci.Request` to store ABCI++ requests of any type.

Reviewer: ```suggestion
- Returns the array of slices where each slice contains the set of `abci.Request`s logged on that node. Every time a crash happens, a new array element (new slice `[]*abci.Request`) will be created. We know a crash has happened because we log "Application started" every time the application starts. Specifically, we added this log inside `NewApplication()` function in `test/e2e/app/app.go` file. In the end, `fetchABCIRequests()` will return just one slice if the node did not experience any crashes and $n+1$ slices if there were $n$ crashes. The benefit of logging the requests in the previously described way is that now we can use `[]*abci.Request` to store ABCI++ requests of any type.
```
it's just a suggestion

ConsensusRounds : ConsensusRound | ConsensusRound ConsensusRounds ;
ConsensusRound : Proposer | NonProposer ; 

Proposer : PrepareProposal | PrepareProposal ProcessProposal ; 
NonProposer: ProcessProposal ;

InitChain : "init_chain" ;
FinalizeBlock : "finalize_block" ; 
Commit : "commit" ;
OfferSnapshot : "offer_snapshot" ;
ApplyChunk : "apply_snapshot_chunk" ; 
PrepareProposal : "prepare_proposal" ; 
ProcessProposal : "process_proposal" ;
 
 ```

If you compare this grammar with the original one, you will notice that, in addition to vote extensions,  
`Info` is removed. The reason is that, as explained in the section [CometBFT's expected behaviour](../../spec/abci/abci%2B%2B_comet_expected_behavior.md#valid-method-call-sequences), one of the 
purposes of the `Info` method is being part of the RPC handling from an external 
client. Since this can happen at any time, it complicates the 
grammar.  
This is not true in other cases, but since the Application does 
not know why the `Info` is called, we removed 
it totally from the new grammar. The Application is still logging the `Info` 
call, but a specific test would need to be written to check whether it happens

> in the right moment. 

Reviewer: ```suggestion
at the right moment. 
```

	} else {
		errors = g.verifyRecovery(execution)
	}
	if errors == nil {
		return true, nil
	}
	return false, fmt.Errorf("%v\nFull execution:\n%v", g.combineErrors(errors, g.cfg.NumberOfErrorsToShow), g.addHeightNumbersToTheExecution(execution))
}
```

It receives a set of ABCI++ requests and a flag saying whether they represent a `CleanStart` execution or not and does the following things:
- Checks if the execution is an empty execution. 
- Filter the requests by calling the method `filterRequests()`. This method will remove all the requests from the set that are not supported by the current version of the grammar. In addition, it will filter the last height by removing all ABCI++ requests after the 
last `Commit`. The function `fetchABCIRequests()` can be called in the middle of the height. As a result, the last height may be incomplete and 
classified as invalid, even if that is not the reality. The simple example here is that the last 
request fetched via `fetchABCIRequests()` is `FinalizeBlock`; however, `Commit` happens after 
`fetchABCIRequests()` was invoked. Consequently, the execution
will be considered as faulty because `Commit` is missing, even though the `Commit` 

> will happen after. This is why if the execution consists of only one incomplete height and function `filterRequests()` returns an empty set of requests, the `Verify()` method considers this execution as valid and returns `true`. 

Reviewer: If could be good to mention here that this mechanism effectively turns the original grammar, which is infinite (omega grammar) into a finite one. In terms of the grammar, we change multiplier `(inf)` by `*`.

Anyway, this is a nice-to-have. Up to you if you want to explain it here.

Author: Addressed. Not precisely here, but in the part where we explain the difference between the new and original grammar.

*Note:* The `gogll` parser can return many errors because it returns an error at every point at which the parser fails to parse
a grammar production. Usually, the error of interest is the one that has 
parsed the largest number of tokens. This is why, by default, we are printing only the last error; however, this can be configured with the `NumberOfErrorsToShow` field of `GrammarChecker`'s config.

**Changing the grammar**

Any modification to the grammar (`test/e2e/pkg/grammar/clean-start/abci_grammar_clean_start.md` or `test/e2e/pkg/grammar/recovery/abci_grammar_recovery.md`) requires generating a new parser and lexer. This is done by 
going to the `test/e2e/` directory and running:

```bash 
make grammar
``` 
Notice here that you need to have `gogll` installed 
on your machine to run the make successfully. If this is not the case, you can install it with the following command: 

```bash 
go get github.com/goccmack/gogll/v3
```  

> 

Reviewer: ```suggestion
Make sure you commit any changes to the auto-generated code together with the changes to the grammar.

```

### Suporting additional ABCI++ requests

Here we present all the steps we need to do if we want to support other 
ABCI++ requests in the future: 

- The application needs to log the new request in the same way as we do now.
- We should include the new request to the grammar and generate a new parser and lexer.  
- We should add new requests to the list of supported requests. Namely, we should modify the function `isSupportedByGrammar()` in `test/e2e/pkg/grammar/checker.go` to return `true` for the new type of requests.

## Status

Implemented.

To-do list:

> - adding the CI workflow to check if make grammar is executed. 

Reviewer: ```suggestion
- adding the CI workflow to check if make grammar is executed. 
- extend this ADR (and implementation) to support ABCI 2.0 (i.e., ABCI calls related to vote extensions)
- in the future, we might consider whether the logging (actually, tracing) should be done on the e2e application side, or on CometBFT side, so this infra can be reused for MBT-like activities)
```

call, but a specific test would need to be written to check whether it happens
in the right moment. 

> Moreover, both grammars, the original and the new, represent the node's expected behaviour from the fresh beginning (`CleanStart`) or after the crash (`Recovery`).

Reviewer: ```suggestion
Moreover, both grammars, the original and the new, represent the node's expected behaviour from the fresh beginning (`CleanStart`) or after a crash (`Recovery`).
```

will be replaced with a corresponding terminal. For example, request `r` of type `abci.Request_PrepareProposal` is replaced with the string `prepare_proposal`, the first part of `r`'s string representation. 
- Checks if the resulting string with terminals respects the grammar by calling the 
appropriate function (`verifyCleanStart()` or `verifyRecovery()`) depending on the execution type. The implementations of both functions are the same; they just use different parsers and lexers. 
- Returns true if the execution is valid and an error if that's not the case. An example of an error is below. 

```
FAIL: TestABCIGrammar/full02 (8.76s)
        abci_test.go:24: ABCI grammar verification failed: The error: "Invalid clean-start execution: parser was expecting one of [init_chain], got [offer_snapshot] instead." has occured at height 0.
            
            Full execution:
            0: offer_snapshot apply_snapshot_chunk finalize_block commit
            1: finalize_block commit
            2: finalize_block commit
            3: finalize_block commit
			...
```

> The error shown above reports an invalid `CleanStart` execution. Moreover, it says why it is considered invalid (`init_chain` was missing) and the height of the error. Notice here that the height in the case of `CleanStart` execution corresponds to the actual consensus height, while for the `Recovery` execution, height 0 represents the first height after the crash. Lastly, after the error, the full execution, one height per line, is printed. This part may be optional and handled with the config flag, but we left it like this for now. 

Reviewer: ```suggestion
The error shown above reports an invalid `CleanStart` execution. Moreover, it says why it is considered invalid (`init_chain` was missing) and the height of the error. Notice here that the height in the case of `CleanStart` execution corresponds to the actual consensus height, while for the `Recovery` execution, height 0 represents the first height after the crash. Lastly, after the error, the full execution, one height per line, is printed. This part may be optional and handled with a config flag, but we left it like this for now. 
```
